File: bindings/java/src/main/java/io/github/endeavoury/ratatoskr/Ratatoskr.java
package io.github.endeavoury.ratatoskr;

import io.github.endeavoury.ratatoskr.internal.NativeBindings;

/** Entry point and native version information for Ratatoskr. */
public final class Ratatoskr {
    private Ratatoskr() {}

    /** Returns a thread-safe DNS client using system resolver defaults. */
    public static DnsClient dns() {
        return new DnsClient();
    }

    /** Returns the loaded native ABI version. */
    public static int abiVersion() {
        return NativeBindings.abiVersion();
    }

    /** Returns the loaded native product version. */
    public static String version() {
        return NativeBindings.versionMajor() + "." + NativeBindings.versionMinor()
            + "." + NativeBindings.versionPatch();
    }
}

File: bindings/kotlin/src/main/kotlin/io/github/endeavoury/ratatoskr/kotlin/Ratatoskr.kt
package io.github.endeavoury.ratatoskr.kotlin

import io.github.endeavoury.ratatoskr.DnsClient as JavaDnsClient
import io.github.endeavoury.ratatoskr.DnsQueryOptions as JavaDnsQueryOptions
import io.github.endeavoury.ratatoskr.DnsRecord as JavaDnsRecord
import io.github.endeavoury.ratatoskr.DnsRecordType
import io.github.endeavoury.ratatoskr.DnsResult as JavaDnsResult
import io.github.endeavoury.ratatoskr.Ratatoskr as JavaRatatoskr

typealias DnsRecord = JavaDnsRecord
typealias DnsResult = JavaDnsResult
typealias DnsQueryOptions = JavaDnsQueryOptions

object Ratatoskr {
    val abiVersion: Int get() = JavaRatatoskr.abiVersion()
    val version: String get() = JavaRatatoskr.version()
    fun dns(options: DnsQueryOptions = DnsQueryOptions.defaults()) = DnsClient(options)
}

class DnsClient(options: DnsQueryOptions = DnsQueryOptions.defaults()) {
    private val delegate = JavaDnsClient(options)
    fun query(name: String, type: DnsRecordType = DnsRecordType.A): DnsResult = delegate.query(name, type)
}
